File: app/src/main/java/wangdaye/com/geometricweather/background/polling/basic/UpdateService.java
package wangdaye.com.geometricweather.background.polling.basic;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.List;

import javax.inject.Inject;

import wangdaye.com.geometricweather.background.polling.PollingUpdateHelper;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.utils.helpters.AsyncHelper;
import wangdaye.com.geometricweather.common.utils.helpters.ShortcutsHelper;
import wangdaye.com.geometricweather.db.DatabaseHelper;
import wangdaye.com.geometricweather.location.LocationHelper;
import wangdaye.com.geometricweather.remoteviews.NotificationHelper;
import wangdaye.com.geometricweather.weather.WeatherHelper;

/**
 * Update service.
 * */

public abstract class UpdateService extends Service
        implements PollingUpdateHelper.OnPollingUpdateListener {

    private PollingUpdateHelper mPollingHelper;
    @Inject LocationHelper mLocationHelper;
    @Inject WeatherHelper mWeatherHelper;
    private List<Location> mLocationList;
    private AsyncHelper.Controller mDelayController;
    private boolean mFailed;

    @Override
    public void onCreate() {
        super.onCreate();

        mFailed = false;

        mLocationList = DatabaseHelper.getInstance(this).readLocationList();

        mPollingHelper = new PollingUpdateHelper(
                this, mLocationHelper, mWeatherHelper, mLocationList);
        mPollingHelper.setOnPollingUpdateListener(this);
        mPollingHelper.pollingUpdate();

        mDelayController = AsyncHelper.delayRunOnIO(() -> stopService(true), 30 * 1000);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mDelayController != null) {
            mDelayController.cancel();
            mDelayController = null;
        }
        if (mPollingHelper != null) {
            mPollingHelper.setOnPollingUpdateListener(null);
            mPollingHelper.cancel();
            mPollingHelper = null;
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    // control.

    public abstract void updateView(Context context, Location location);

    public abstract void updateView(Context context, List<Location> locationList);

    public abstract void handlePollingResult(boolean updateSucceed);

    public void stopService(boolean updateFailed) {
        handlePollingResult(updateFailed);
        stopSelf();
    }

    // interface.

    // on polling update listener.

    @Override
    public void onUpdateCompleted(@NonNull Location location, @Nullable Weather old,
                                  boolean succeed, int index, int total) {
        for (int i = 0; i < mLocationList.size(); i ++) {
            if (mLocationList.get(i).equals(location)) {
                mLocationList.set(i, location);
                if (i == 0) {
                    updateView(this, location);
                    if (succeed) {
                        NotificationHelper.checkAndSendAlert(this, location, old);
                        NotificationHelper.checkAndSendPrecipitationForecast(this, location, old);
                    } else {
                        mFailed = true;
                    }
                }
                return;
            }
        }
    }

    @Override
    public void onPollingCompleted() {
        updateView(this, mLocationList);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
            ShortcutsHelper.refreshShortcutsInNewThread(this, mLocationList);
        }
        stopService(mFailed);
    }
}

File: app/src/main/java/org/breezyweather/background/polling/services/basic/UpdateService.kt
package org.breezyweather.background.polling.services.basic

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import org.breezyweather.background.polling.PollingUpdateHelper
import org.breezyweather.background.polling.PollingUpdateHelper.OnPollingUpdateListener
import org.breezyweather.common.basic.models.Location
import org.breezyweather.common.basic.models.weather.Weather
import org.breezyweather.common.utils.helpers.AsyncHelper
import org.breezyweather.common.utils.helpers.ShortcutsHelper
import org.breezyweather.location.LocationHelper
import org.breezyweather.remoteviews.NotificationHelper
import org.breezyweather.weather.WeatherHelper
import javax.inject.Inject

abstract class UpdateService : Service(), OnPollingUpdateListener {
    private var mPollingHelper: PollingUpdateHelper? = null

    var mLocationHelper: LocationHelper? = null
        @Inject set

    var mWeatherHelper: WeatherHelper? = null
        @Inject set

    private var mDelayController: AsyncHelper.Controller? = null
    private var mFailed = false

    override fun onCreate() {
        super.onCreate()
        mFailed = false
        mPollingHelper = PollingUpdateHelper(this, mLocationHelper!!, mWeatherHelper!!).also {
            it.setOnPollingUpdateListener(this)
            it.pollingUpdate()
        }
        mDelayController = AsyncHelper.delayRunOnIO({ stopService(true) }, (30 * 1000).toLong())
    }

    override fun onDestroy() {
        super.onDestroy()
        mDelayController?.let {
            it.cancel()
            mDelayController = null
        }
        mPollingHelper?.let {
            it.setOnPollingUpdateListener(null)
            it.cancel()
            mPollingHelper = null
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    // control.
    abstract fun updateView(context: Context, location: Location)
    abstract fun updateView(context: Context, locationList: List<Location>)
    abstract fun handlePollingResult(updateSucceed: Boolean)
    open fun stopService(updateFailed: Boolean) {
        handlePollingResult(updateFailed)
        stopSelf()
    }

    // interface.
    // on polling update listener.
    override fun onUpdateCompleted(
        location: Location, old: Weather?,
        succeed: Boolean, index: Int, total: Int
    ) {
        if (index == 0) {
            updateView(this, location)
            if (succeed) {
                NotificationHelper.checkAndSendAlert(this, location, old)
                NotificationHelper.checkAndSendPrecipitationForecast(this, location)
            } else {
                mFailed = true
            }
        }
    }

    override fun onPollingCompleted(locationList: List<Location>) {
        updateView(this, locationList)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
            ShortcutsHelper.refreshShortcutsInNewThread(this, locationList)
        }
        stopService(mFailed)
    }
}
